src/main/java/com/laytonsmith/core/functions/Threading.java

```diff
@@ -329,4 +336,137 @@ public Version since() {
 
 	}
 
+	@api
+	@noboilerplate
+	@seealso({x_new_thread.class})
+	public static class _synchronized extends AbstractFunction {
+		private static final HashMap<Object, Integer> syncObjectMap = new HashMap<Object, Integer>();
+		
+		@Override
+		public Class<? extends CREThrowable>[] thrown() {
+			return new Class[]{CRENullPointerException.class};
+		}
+		
+		@Override
+		public boolean isRestricted() {
+			return true;
+		}
+		
+		@Override
+		public Boolean runAsync() {
+			return false;
+		}
+		
+		@Override
+		public boolean preResolveVariables() {
+			return false;
+		}
+		
+		@Override
+		public boolean useSpecialExec() {
+			return true;
+		}
+		
+		@Override
+		public Construct execs(Target t, Environment env, Script parent, ParseTree... nodes) {
+			
+			// Get the sync object tree and the code to synchronize.
+			ParseTree syncObjectTree = nodes[0];
+			ParseTree code = nodes[1];
+			
+			// Get the sync object (CArray or String value of the Construct).
+			Construct cSyncObject = parent.seval(syncObjectTree, env);
+			if(cSyncObject instanceof CNull) {
+				throw new CRENullPointerException("Synchronization object may not be null in " + getName() + "().", t);
+			}
+			Object syncObject;
+			if(cSyncObject instanceof CArray) {
+				syncObject = cSyncObject;
+			} else {
+				syncObject = cSyncObject.val();
+			}
+			
+			// Add String sync objects to the map to be able to synchronize by value.
+			if(syncObject instanceof String) {
+				synchronized(syncObjectMap) {
+					searchLabel: {
+						for(Entry<Object, Integer> entry : syncObjectMap.entrySet()) {
+							Object key = entry.getKey();
+							if(key instanceof String && key.equals(syncObject)) {
+								syncObject = key; // Get reference, value of this assign is the same.
+								entry.setValue(entry.getValue() + 1);
+								break searchLabel;
+							}
+						}
+						syncObjectMap.put(syncObject, 1);
+					}
+				}
+			}
+			
+			// Evaluate the code, synchronized by the passed sync object.
+			Construct ret;
+			try {
+				synchronized(syncObject) {
+					ret = parent.seval(code, env);
+				}
+			} catch(RuntimeException e) {
+				throw e;
+			} finally {
+				
+				// Remove 1 from the call count or remove the sync object from the map if it was a sync-by-value.
+				if(syncObject instanceof String) {
+					synchronized(syncObjectMap) {
+						int count = syncObjectMap.get(syncObject); // This should never return null.
+						if(count <= 1) {
+							syncObjectMap.remove(syncObject);
+						} else {
+							for(Entry<Object, Integer> entry : syncObjectMap.entrySet()) {
+								if(entry.getKey() == syncObject) { // Equals by reference.
+									entry.setValue(count - 1);
+									break;
+								}
+							}
+						}
+					}
+				}
+			}
+			
+			// Return the Construct received from evaluating the passed code block.
+			return ret;
```

Comment:
I left it in because I thought this syntax could be nice to use since it matches up with if():
`@var = synchronized(@array) { if(array_size(@array) > 1) { @array[0] } else { null } }`
I guess this could be seen as a feature or as bad programming, what do you prefer?